src/Microsoft.Net.Sdk.AnalyzerRedirecting/SdkAnalyzerAssemblyRedirector.cs

```diff
@@ -0,0 +1,132 @@
+        _analyzerMap = new(CreateAnalyzerMap);
+    }
+
+    /// <summary>
+    /// Map from analyzer assembly name (file name without extension) to a list of matching analyzers.
+    /// </summary>
+    private ImmutableDictionary<string, List<AnalyzerInfo>> AnalyzerMap => _analyzerMap.Value;
+
+    private ImmutableDictionary<string, List<AnalyzerInfo>> CreateAnalyzerMap()
+    {
+        var builder = ImmutableDictionary.CreateBuilder<string, List<AnalyzerInfo>>(StringComparer.OrdinalIgnoreCase);
+
+        // Expects layout like:
+        // VsInstallDir\SDK\RuntimeAnalyzers\WindowsDesktopAnalyzers\8.0.8\analyzers\dotnet\System.Windows.Forms.Analyzers.dll
+        //                                   ~~~~~~~~~~~~~~~~~~~~~~~                                                           = topLevelDirectory
+        //                                                           ~~~~~                                                     = versionDirectory
+        //                                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ = analyzerPath
+
+        foreach (string topLevelDirectory in Directory.EnumerateDirectories(_insertedAnalyzersDirectory))
+        {
+            foreach (string versionDirectory in Directory.EnumerateDirectories(topLevelDirectory))
+            {
+                foreach (string analyzerPath in Directory.EnumerateFiles(versionDirectory, "*.dll", SearchOption.AllDirectories))
+                {
+                    if (!analyzerPath.StartsWith(versionDirectory, StringComparison.OrdinalIgnoreCase))
+                    {
+                        continue;
+                    }
+
+                    string version = Path.GetFileName(versionDirectory);
+                    string analyzerName = Path.GetFileNameWithoutExtension(analyzerPath);
+                    string pathSuffix = analyzerPath.Substring(versionDirectory.Length + 1 /* slash */);
+                    pathSuffix = Path.GetDirectoryName(pathSuffix);
+
+                    AnalyzerInfo analyzer = new() { FullPath = analyzerPath, ProductVersion = version, PathSuffix = pathSuffix };
+
+                    if (builder.TryGetValue(analyzerName, out var existing))
+                    {
+                        existing.Add(analyzer);
+                    }
+                    else
+                    {
+                        builder.Add(analyzerName, [analyzer]);
+                    }
+                }
+            }
+        }
+
+        return builder.ToImmutable();
+    }
+
+    public string? RedirectPath(string fullPath)
+    {
+        if (AnalyzerMap.TryGetValue(Path.GetFileNameWithoutExtension(fullPath), out var analyzers))
+        {
+            foreach (AnalyzerInfo analyzer in analyzers)
+            {
+                var directoryPath = Path.GetDirectoryName(fullPath);
+                if (directoryPath.EndsWith(analyzer.PathSuffix, StringComparison.OrdinalIgnoreCase) &&
```

Comment:
Can we guarantee that the path passed in here will be normalized in the same way that `pathSuffix` is?  If `fullPath` is coming from a user-controlled item in MSBuild I could imagine it has all sorts of nonsense in it (relative paths, extra path separators, alternate path separators, etc).

Reply:
Good point. But since this is targeting .net framework, `Path.GetDirectoryName` normalizes the path. I should add a comment.

Reply:
You could also explicitly [normalize](https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#path-normalization) it with `Path.GetFullPath`